lib/linked_list.rb

```diff
@@ -19,54 +19,125 @@ def initialize
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
     def add_first(value)
-      raise NotImplementedError
+      new_node = Node.new(value, @head)
+      @head = new_node 
     end
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
     def search(value)
-      raise NotImplementedError
+      return false if @head == nil
+      
+      current = @head
+      found = false
+      until current == nil
+        if current.data == value
+          found = true
+          break
+        end 
+
+        current = current.next
+      end
+      return found
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max
-      raise NotImplementedError
+      return nil if @head == nil
+      
+      current = @head
+      max = current.data
+      
+      until current == nil
+        if current.data > max
+          max = current.data
+        end
+        current = current.next
+      end
+      return max
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
     def find_min
-      raise NotImplementedError
+      return nil if @head == nil
+      
+      current = @head
+      min = 0

```

Comment:
What if all the elements are negative numbers?

```suggestion
      min = current.data
```